src/features/loading/status-announcer.ts

```diff
@@ -0,0 +1,181 @@
+
+  constructor() {
+    this.statusElement.setAttribute('role', 'status');
+    this.statusElement.style.position = 'absolute';
+    this.statusElement.style.color = 'transparent';
+    this.statusElement.style.pointerEvents = 'none';
+  }
+
+  /**
+   * Register a <model-viewer> element with the announcer. If it is not yet
+   * loaded, its loading status will be tracked by the announcer.
+   */
+  registerInstance(modelViewer: ModelViewerElementBase) {
+    if (this.registeredInstanceStatuses.has(modelViewer)) {
+      return;
+    }
+
+    let instanceWasUnregistered = () => {};
+    const loadShouldBeMeasured =
+        modelViewer.loaded === false && !!(modelViewer as any).src;
+    const loadAttemptCompletes = loadShouldBeMeasured ?
+        new Promise((resolve) => {
+          const resolveHandler = () => {
+            resolve();
+
+            modelViewer.removeEventListener('load', resolveHandler);
+            modelViewer.removeEventListener('error', resolveHandler);
+          };
+
+          modelViewer.addEventListener('load', resolveHandler);
+          modelViewer.addEventListener('error', resolveHandler);
+
+          instanceWasUnregistered = resolveHandler;
+        }) :
+        Promise.resolve();
+
+    this.registeredInstanceStatuses.set(
+        modelViewer, {instanceWasUnregistered, loadAttemptCompletes});
+
+    this.loadingPromises.push(loadAttemptCompletes);
+
+    if (this.modelViewerStatusInstance == null) {
+      this.modelViewerStatusInstance = modelViewer;
+    }
+  }
+
+  /**
+   * Unregister a <model-viewer> element with the announcer. Its loading status
+   * will no longer be tracked by the announcer.
+   */
+  unregisterInstance(modelViewer: ModelViewerElementBase) {
+    if (!this.registeredInstanceStatuses.has(modelViewer)) {
+      return;
+    }
+
+    const statuses = this.registeredInstanceStatuses;
+    const instanceStatus = statuses.get(modelViewer)!;
+    statuses.delete(modelViewer);
+    instanceStatus.instanceWasUnregistered();
+
+    if (this.modelViewerStatusInstance === modelViewer) {
+      this.modelViewerStatusInstance = statuses.size > 0 ?
+          getFirstMapKey<ModelViewerElementBase, InstanceLoadingStatus>(
+              statuses) :
+          null;
+    }
+  }
+
+  protected get modelViewerStatusInstance(): ModelViewerElementBase|null {
+    return this[$modelViewerStatusInstance];
+  }
+
+  protected set modelViewerStatusInstance(value: ModelViewerElementBase|null) {
+    const currentInstance = this[$modelViewerStatusInstance];
+    if (currentInstance === value) {
+      return;
+    }
+
+    const {statusElement} = this;
+
+    if (value != null && value.shadowRoot != null) {
+      value.shadowRoot.appendChild(statusElement);
+    } else if (statusElement.parentNode != null) {
+      statusElement.parentNode.removeChild(statusElement);
+    }
+
+    this[$modelViewerStatusInstance] = value;
+    this[$updateStatus]();
+  }
+
+  protected async updateStatus() {
```

Comment:
IIUC, setting `modelViewerStatusInstance` starts the debounced polling, which only happens when an element is registered, and it's the first element to do so (when `modelViewStatusInstance` has not yet been created), and when an element is disconnected.

What of a scenario where `n` models are loaded, and later, a new model has been added, calling `registerInstance`, and then not starting the polling because `modelViewerStatusInstance` already exists? This could already be covered and I'm missing a piece, it looks like the last test *could* cover this, but I suspect both elements are registered immediately, while the initial debounce cycle is still running

Reply:
My thoughts were that there were two important scenarios to cover:

 1. First load of the page, where we will pick up N elements and then conclude when they have loaded (including lazily added ones in the mean time)
 2. The case where the site is a SPA, so a page navigation implies a new battery of elements might be added, which means we should do the announcement flow over again

Notably this leaves out a third case (as I think you are pointing out):

 3. A `<model-viewer>` is added after the final announcement, but before any of the previous `<model-viewer>` elements have been removed

This was done deliberately. My thinking here is that the risk of spammy loading announcements was a worse outcome than not saying anything about a sufficiently late-added `<model-viewer>`.